File: Depth-First Search/Combinations/String/LET_no131_Palindrome Partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> temp;
        dfs(res, s, temp, 0);
        return res;
    }
    
    void dfs(vector<vector<string>>& res, string& s, vector<string>& temp, int start) {
        if (start == s.size()) {
            res.push_back(temp);
            return;
        }
        
        for (int i = start; i < s.size(); i++) {
            if (isPalindrome(s, start, i)) {
                temp.push_back(s.substr(start, i - start + 1));
                dfs(res, s, temp, i + 1);
                temp.pop_back();
            }
        }
    }
    
    bool isPalindrome(string& s, int left, int right) {
        while (left <= right) {
            if (s[left++] != s[right--])
                return false;
        }
        return true;
    }
};

File: Depth-First Search/Combinations/String/LET_no131_Palindrome Partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        self.dfs(s, 0, [], res)
        return res
    
    def dfs(self, s, start, temp, res):
        if start == len(s):
            res.append(temp[:])
            return
        
        for i in range(start, len(s)):
            if self.isPalindrome(s, start, i):
                temp.append(s[start:i+1])
                self.dfs(s, i + 1, temp, res)
                temp.pop()
    
    def isPalindrome(self, s, left, right):
        while left < right and s[left] == s[right]:
            left += 1
            right -= 1
        return left >= right